shim.js

```diff
@@ -14,167 +14,172 @@
-// module.
-if (needExecArgv.length) {
-  var pexec = process.execArgv
-  if (JSON.stringify(pexec) !== JSON.stringify(needExecArgv)) {
-    debug('need execArgv for this', pexec, '=>', needExecArgv)
-    var spawn = require('child_process').spawn
-    var sargs = pexec.concat(needExecArgv).concat(process.argv.slice(1))
-    foregroundChild(node, sargs)
+  debug('after argv parse hasMain=%j', hasMain)
+
+  if (hasMain > 2) {
+    // if the main file is above #2, then it means that there
+    // was a --exec_arg *before* it.  This means that we need
+    // to slice everything from 2 to hasMain, and pass that
+    // directly to node.  This also splices out the user-supplied
+    // execArgv from the argv.
+    const addExecArgv = process.argv.splice(2, hasMain - 2)
+    needExecArgv.push.apply(needExecArgv, addExecArgv)
+  }
+
+  if (!hasMain) {
+    // we got loaded by mistake for a `node -pe script` or something.
+    const args = process.execArgv.concat(needExecArgv, process.argv.slice(2))
+    debug('no main file!', args)
+    foregroundChild(node, args)
     return
   }
-}
-
-// At this point, we've verified that we got the correct execArgv,
-// and that we have a main file, and that the main file is sitting at
-// argv[2].  Splice this shim off the list so it looks like the main.
-var spliceArgs = [1, 1].concat(argv)
-process.argv.splice.apply(process.argv, spliceArgs)
-
-// Unwrap the PATH environment var so that we're not mucking
-// with the environment.  It'll get re-added if they spawn anything
-var isWindows = (
-  process.platform === 'win32' ||
-  process.env.OSTYPE === 'cygwin' ||
-  process.env.OSTYPE === 'msys'
-)
-
-if (isWindows) {
-  for (var i in process.env) {
-    if (i.match(/^path$/i)) {
-      process.env[i] = process.env[i].replace(__dirname + ';', '')
+
+  // If there are execArgv, and they're not the same as how this module
+  // was executed, then we need to inject those.  This is for stuff like
+  // --harmony or --use_strict that needs to be *before* the main
+  // module.
+  if (needExecArgv.length) {
+    const pexec = process.execArgv
+    if (JSON.stringify(pexec) !== JSON.stringify(needExecArgv)) {
+      debug('need execArgv for this', pexec, '=>', needExecArgv)
+      const spawn = require('child_process').spawn
+      const sargs = pexec.concat(needExecArgv).concat(process.argv.slice(1))
+      foregroundChild(node, sargs)
+      return
+    }
+  }
+
+  // At this point, we've verified that we got the correct execArgv,
+  // and that we have a main file, and that the main file is sitting at
+  // argv[2].  Splice this shim off the list so it looks like the main.
+  const spliceArgs = [1, 1].concat(argv)
+  process.argv.splice.apply(process.argv, spliceArgs)
+
+  // Unwrap the PATH environment var so that we're not mucking
+  // with the environment.  It'll get re-added if they spawn anything
+  const isWindows = (
+    // TODO: Use `lib/is-windows`
```

Comment:
Better we can just add an `isWindows` boolean value to the settings.json.

Reply:
I forgot to remove this comment, but I definitely agree.
My plan is to move the environment manipulation functions to their own module (they're also used in the "host" part) and pass this module to the shim. The shim won't have to worry about the platform-dependent logic, it is encapsulated by the env-vars module.

<https://github.com/demurgos/spawn-wrap/blob/5c0c758b0ba1ac93b3705570ad7dadc0fb046355/src/lib/path-env-var.ts>
<https://github.com/demurgos/spawn-wrap/blob/next/src/lib/templates/shim-template.ts#L22>